libfwupdplugin/fu-optionrom-device.c

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2015 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ */
+
+#include "config.h"
+
+#include "fu-optionrom-device.h"
+
+G_DEFINE_TYPE(FuOptionromDevice, fu_optionrom_device, FU_TYPE_PCI_DEVICE)
+
+static gboolean
+fu_optionrom_device_probe(FuDevice *device, GError **error)
+{
+	FuOptionromDevice *self = FU_OPTIONROM_DEVICE(device);
+	gboolean rom_exists = FALSE;
+	g_autofree gchar *rom_fn = NULL;
+
+	/* does the device even have ROM? */
+	rom_fn = g_build_filename(fu_udev_device_get_sysfs_path(FU_UDEV_DEVICE(self)), "rom", NULL);
+	if (!fu_device_query_file_exists(device, rom_fn, &rom_exists, error))
+		return FALSE;
+	if (rom_exists)
+		fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_CAN_VERIFY_IMAGE);
+	return TRUE;
+}
+
+static GBytes *
+fu_optionrom_device_dump_firmware(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuOptionromDevice *self = FU_OPTIONROM_DEVICE(device);
+	guint number_reads = 0;
+	g_autofree gchar *rom_fn = NULL;
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+	g_autoptr(GError) error_local = NULL;
+	g_autoptr(GFile) file = NULL;
+	g_autoptr(GInputStream) stream = NULL;
+
+	/* sanity check */
+	if (!fu_device_has_flag(device, FWUPD_DEVICE_FLAG_CAN_VERIFY_IMAGE)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "unable to read firmware from device, 'rom' does not exist");
+		return NULL;
+	}
+
+	/* open file */
+	rom_fn = g_build_filename(fu_udev_device_get_sysfs_path(FU_UDEV_DEVICE(self)), "rom", NULL);
+	file = g_file_new_for_path(rom_fn);
+	stream = G_INPUT_STREAM(g_file_read(file, NULL, &error_local));
+	if (stream == NULL) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_AUTH_FAILED,
+				    error_local->message);
+		return NULL;
+	}
+
+	/* we have to enable the read for devices */
+	if (g_str_has_prefix(rom_fn, "/sys")) {
```

Comment:
won't `fu_udev_device_get_sysfs_path` always give you something with `/sys`?  Or is this with the lens of emulation and tests?

Reply:
Exactly the latter.